File: binding/include/libqquaternion.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qquaternion.html
pub const qquaternion = struct {

    /// New constructs a new QQuaternion object.
    pub fn New(other: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_new(@ptrCast(other));
    }


    /// New2 constructs a new QQuaternion object.
    pub fn New2(other: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_new2(@ptrCast(other));
    }


    /// New3 constructs a new QQuaternion object.
    pub fn New3() QtC.QQuaternion {
        return qtc.QQuaternion_new3();
    }


    /// New4 constructs a new QQuaternion object.
    pub fn New4(param1: i32) QtC.QQuaternion {
        return qtc.QQuaternion_new4(@intCast(param1));
    }


    /// New5 constructs a new QQuaternion object.
    pub fn New5(scalar: f32, xpos: f32, ypos: f32, zpos: f32) QtC.QQuaternion {
        return qtc.QQuaternion_new5(scalar, xpos, ypos, zpos);
    }


    /// New6 constructs a new QQuaternion object.
    pub fn New6(scalar: f32, vector: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_new6(scalar, @ptrCast(vector));
    }


    /// New7 constructs a new QQuaternion object.
    pub fn New7(vector: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_new7(@ptrCast(vector));
    }


    /// New8 constructs a new QQuaternion object.
    pub fn New8(param1: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_new8(@ptrCast(param1));
    }

    /// CopyAssign shallow copies `other` into `self`.
    pub fn CopyAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QQuaternion_CopyAssign(@ptrCast(self), @ptrCast(other));
    }

    /// MoveAssign moves `other` into `self` and invalidates `other`.
    pub fn MoveAssign(self: ?*anyopaque, other: ?*anyopaque) void {
        qtc.QQuaternion_MoveAssign(@ptrCast(self), @ptrCast(other));
    }


    pub fn IsNull(self: ?*anyopaque, ) bool {
        return qtc.QQuaternion_IsNull(@ptrCast(self));
    }

    pub fn IsIdentity(self: ?*anyopaque, ) bool {
        return qtc.QQuaternion_IsIdentity(@ptrCast(self));
    }

    pub fn Vector(self: ?*anyopaque, ) QtC.QVector3D {
        return qtc.QQuaternion_Vector(@ptrCast(self));
    }

    pub fn SetVector(self: ?*anyopaque, vector: ?*anyopaque) void {
        qtc.QQuaternion_SetVector(@ptrCast(self), @ptrCast(vector));
    }

    pub fn SetVector2(self: ?*anyopaque, x: f32, y: f32, z: f32) void {
        qtc.QQuaternion_SetVector2(@ptrCast(self), x, y, z);
    }

    pub fn X(self: ?*anyopaque, ) f32 {
        return qtc.QQuaternion_X(@ptrCast(self));
    }

    pub fn Y(self: ?*anyopaque, ) f32 {
        return qtc.QQuaternion_Y(@ptrCast(self));
    }

    pub fn Z(self: ?*anyopaque, ) f32 {
        return qtc.QQuaternion_Z(@ptrCast(self));
    }

    pub fn Scalar(self: ?*anyopaque, ) f32 {
        return qtc.QQuaternion_Scalar(@ptrCast(self));
    }

    pub fn SetX(self: ?*anyopaque, x: f32) void {
        qtc.QQuaternion_SetX(@ptrCast(self), x);
    }

    pub fn SetY(self: ?*anyopaque, y: f32) void {
        qtc.QQuaternion_SetY(@ptrCast(self), y);
    }

    pub fn SetZ(self: ?*anyopaque, z: f32) void {
        qtc.QQuaternion_SetZ(@ptrCast(self), z);
    }

    pub fn SetScalar(self: ?*anyopaque, scalar: f32) void {
        qtc.QQuaternion_SetScalar(@ptrCast(self), scalar);
    }

    pub fn DotProduct(q1: ?*anyopaque, q2: ?*anyopaque) f32 {
        return qtc.QQuaternion_DotProduct(@ptrCast(q1), @ptrCast(q2));
    }

    pub fn Length(self: ?*anyopaque, ) f32 {
        return qtc.QQuaternion_Length(@ptrCast(self));
    }

    pub fn LengthSquared(self: ?*anyopaque, ) f32 {
        return qtc.QQuaternion_LengthSquared(@ptrCast(self));
    }

    pub fn Normalized(self: ?*anyopaque, ) QtC.QQuaternion {
        return qtc.QQuaternion_Normalized(@ptrCast(self));
    }

    pub fn Normalize(self: ?*anyopaque, ) void {
        qtc.QQuaternion_Normalize(@ptrCast(self));
    }

    pub fn Inverted(self: ?*anyopaque, ) QtC.QQuaternion {
        return qtc.QQuaternion_Inverted(@ptrCast(self));
    }

    pub fn Conjugated(self: ?*anyopaque, ) QtC.QQuaternion {
        return qtc.QQuaternion_Conjugated(@ptrCast(self));
    }

    pub fn RotatedVector(self: ?*anyopaque, vector: ?*anyopaque) QtC.QVector3D {
        return qtc.QQuaternion_RotatedVector(@ptrCast(self), @ptrCast(vector));
    }

    pub fn OperatorPlusAssign(self: ?*anyopaque, quaternion: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_OperatorPlusAssign(@ptrCast(self), @ptrCast(quaternion));
    }

    pub fn OperatorMinusAssign(self: ?*anyopaque, quaternion: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_OperatorMinusAssign(@ptrCast(self), @ptrCast(quaternion));
    }

    pub fn OperatorMultiplyAssign(self: ?*anyopaque, factor: f32) QtC.QQuaternion {
        return qtc.QQuaternion_OperatorMultiplyAssign(@ptrCast(self), factor);
    }

    pub fn OperatorMultiplyAssign2(self: ?*anyopaque, quaternion: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_OperatorMultiplyAssign2(@ptrCast(self), @ptrCast(quaternion));
    }

    pub fn OperatorDivideAssign(self: ?*anyopaque, divisor: f32) QtC.QQuaternion {
        return qtc.QQuaternion_OperatorDivideAssign(@ptrCast(self), divisor);
    }

    pub fn ToVector4D(self: ?*anyopaque, ) QtC.QVector4D {
        return qtc.QQuaternion_ToVector4D(@ptrCast(self));
    }

    pub fn OperatorQVariant(self: ?*anyopaque, ) QtC.QVariant {
        return qtc.QQuaternion_OperatorQVariant(@ptrCast(self));
    }

    pub fn GetAxisAndAngle(self: ?*anyopaque, axis: ?*anyopaque, angle: *f32) void {
        qtc.QQuaternion_GetAxisAndAngle(@ptrCast(self), @ptrCast(axis), @ptrCast(angle));
    }

    pub fn FromAxisAndAngle(axis: ?*anyopaque, angle: f32) QtC.QQuaternion {
        return qtc.QQuaternion_FromAxisAndAngle(@ptrCast(axis), angle);
    }

    pub fn GetAxisAndAngle2(self: ?*anyopaque, x: *f32, y: *f32, z: *f32, angle: *f32) void {
        qtc.QQuaternion_GetAxisAndAngle2(@ptrCast(self), @ptrCast(x), @ptrCast(y), @ptrCast(z), @ptrCast(angle));
    }

    pub fn FromAxisAndAngle2(x: f32, y: f32, z: f32, angle: f32) QtC.QQuaternion {
        return qtc.QQuaternion_FromAxisAndAngle2(x, y, z, angle);
    }

    pub fn ToEulerAngles(self: ?*anyopaque, ) QtC.QVector3D {
        return qtc.QQuaternion_ToEulerAngles(@ptrCast(self));
    }

    pub fn FromEulerAngles(eulerAngles: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_FromEulerAngles(@ptrCast(eulerAngles));
    }

    pub fn GetEulerAngles(self: ?*anyopaque, pitch: *f32, yaw: *f32, roll: *f32) void {
        qtc.QQuaternion_GetEulerAngles(@ptrCast(self), @ptrCast(pitch), @ptrCast(yaw), @ptrCast(roll));
    }

    pub fn FromEulerAngles2(pitch: f32, yaw: f32, roll: f32) QtC.QQuaternion {
        return qtc.QQuaternion_FromEulerAngles2(pitch, yaw, roll);
    }

    pub fn GetAxes(self: ?*anyopaque, xAxis: ?*anyopaque, yAxis: ?*anyopaque, zAxis: ?*anyopaque) void {
        qtc.QQuaternion_GetAxes(@ptrCast(self), @ptrCast(xAxis), @ptrCast(yAxis), @ptrCast(zAxis));
    }

    pub fn FromAxes(xAxis: ?*anyopaque, yAxis: ?*anyopaque, zAxis: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_FromAxes(@ptrCast(xAxis), @ptrCast(yAxis), @ptrCast(zAxis));
    }

    pub fn FromDirection(direction: ?*anyopaque, up: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_FromDirection(@ptrCast(direction), @ptrCast(up));
    }

    pub fn RotationTo(from: ?*anyopaque, to: ?*anyopaque) QtC.QQuaternion {
        return qtc.QQuaternion_RotationTo(@ptrCast(from), @ptrCast(to));
    }

    pub fn Slerp(q1: ?*anyopaque, q2: ?*anyopaque, t: f32) QtC.QQuaternion {
        return qtc.QQuaternion_Slerp(@ptrCast(q1), @ptrCast(q2), t);
    }

    pub fn Nlerp(q1: ?*anyopaque, q2: ?*anyopaque, t: f32) QtC.QQuaternion {
        return qtc.QQuaternion_Nlerp(@ptrCast(q1), @ptrCast(q2), t);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QQuaternion_Delete(@ptrCast(self));
    }
};

File: binding/src/libqquaternion.cpp
#include <QQuaternion>
#include <QVariant>
#include <QVector3D>
#include <QVector4D>
#include <qquaternion.h>
#include "libqquaternion.h"
#include "libqquaternion.hxx"

QQuaternion* QQuaternion_new(const QQuaternion* other) {
	 return new QQuaternion(*other);
}

QQuaternion* QQuaternion_new2(QQuaternion* other) {
	 return new QQuaternion(*other);
}

QQuaternion* QQuaternion_new3() {
	 return new QQuaternion();
}

QQuaternion* QQuaternion_new4(int param1) {
	 return new QQuaternion(static_cast<Qt::Initialization>(param1));
}

QQuaternion* QQuaternion_new5(float scalar, float xpos, float ypos, float zpos) {
	 return new QQuaternion(scalar, xpos, ypos, zpos);
}

QQuaternion* QQuaternion_new6(float scalar, const QVector3D* vector) {
	 return new QQuaternion(scalar, *vector);
}

QQuaternion* QQuaternion_new7(const QVector4D* vector) {
	 return new QQuaternion(*vector);
}

QQuaternion* QQuaternion_new8(const QQuaternion* param1) {
	 return new QQuaternion(*param1);
}

void QQuaternion_CopyAssign(QQuaternion* self, QQuaternion* other) {
    *self = *other;
}

void QQuaternion_MoveAssign(QQuaternion* self, QQuaternion* other) {
    *self = std::move(*other);
}

bool QQuaternion_IsNull(const QQuaternion* self) {
	return self->isNull();
}

bool QQuaternion_IsIdentity(const QQuaternion* self) {
	return self->isIdentity();
}

QVector3D* QQuaternion_Vector(const QQuaternion* self) {
	return new QVector3D(self->vector());
}

void QQuaternion_SetVector(QQuaternion* self, const QVector3D* vector) {
	self->setVector(*vector);
}

void QQuaternion_SetVector2(QQuaternion* self, float x, float y, float z) {
	self->setVector(x, y, z);
}

float QQuaternion_X(const QQuaternion* self) {
	return self->x();
}

float QQuaternion_Y(const QQuaternion* self) {
	return self->y();
}

float QQuaternion_Z(const QQuaternion* self) {
	return self->z();
}

float QQuaternion_Scalar(const QQuaternion* self) {
	return self->scalar();
}

void QQuaternion_SetX(QQuaternion* self, float x) {
	self->setX(x);
}

void QQuaternion_SetY(QQuaternion* self, float y) {
	self->setY(y);
}

void QQuaternion_SetZ(QQuaternion* self, float z) {
	self->setZ(z);
}

void QQuaternion_SetScalar(QQuaternion* self, float scalar) {
	self->setScalar(scalar);
}

float QQuaternion_DotProduct(const QQuaternion* q1, const QQuaternion* q2) {
	return QQuaternion::dotProduct(*q1, *q2);
}

float QQuaternion_Length(const QQuaternion* self) {
	return self->length();
}

float QQuaternion_LengthSquared(const QQuaternion* self) {
	return self->lengthSquared();
}

QQuaternion* QQuaternion_Normalized(const QQuaternion* self) {
	return new QQuaternion(self->normalized());
}

void QQuaternion_Normalize(QQuaternion* self) {
	self->normalize();
}

QQuaternion* QQuaternion_Inverted(const QQuaternion* self) {
	return new QQuaternion(self->inverted());
}

QQuaternion* QQuaternion_Conjugated(const QQuaternion* self) {
	return new QQuaternion(self->conjugated());
}

QVector3D* QQuaternion_RotatedVector(const QQuaternion* self, const QVector3D* vector) {
	return new QVector3D(self->rotatedVector(*vector));
}

QQuaternion* QQuaternion_OperatorPlusAssign(QQuaternion* self, const QQuaternion* quaternion) {
	return new QQuaternion(self->operator+=(*quaternion));
}

QQuaternion* QQuaternion_OperatorMinusAssign(QQuaternion* self, const QQuaternion* quaternion) {
	return new QQuaternion(self->operator-=(*quaternion));
}

QQuaternion* QQuaternion_OperatorMultiplyAssign(QQuaternion* self, float factor) {
	return new QQuaternion(self->operator*=(factor));
}

QQuaternion* QQuaternion_OperatorMultiplyAssign2(QQuaternion* self, const QQuaternion* quaternion) {
	return new QQuaternion(self->operator*=(*quaternion));
}

QQuaternion* QQuaternion_OperatorDivideAssign(QQuaternion* self, float divisor) {
	return new QQuaternion(self->operator/=(divisor));
}

QVector4D* QQuaternion_ToVector4D(const QQuaternion* self) {
	return new QVector4D(self->toVector4D());
}

QVariant* QQuaternion_OperatorQVariant(const QQuaternion* self) {
	return new QVariant(self->operator QVariant());
}

void QQuaternion_GetAxisAndAngle(const QQuaternion* self, QVector3D* axis, float* angle) {
	self->getAxisAndAngle(axis, angle);
}

QQuaternion* QQuaternion_FromAxisAndAngle(const QVector3D* axis, float angle) {
	return new QQuaternion(QQuaternion::fromAxisAndAngle(*axis, angle));
}

void QQuaternion_GetAxisAndAngle2(const QQuaternion* self, float* x, float* y, float* z, float* angle) {
	self->getAxisAndAngle(x, y, z, angle);
}

QQuaternion* QQuaternion_FromAxisAndAngle2(float x, float y, float z, float angle) {
	return new QQuaternion(QQuaternion::fromAxisAndAngle(x, y, z, angle));
}

QVector3D* QQuaternion_ToEulerAngles(const QQuaternion* self) {
	return new QVector3D(self->toEulerAngles());
}

QQuaternion* QQuaternion_FromEulerAngles(const QVector3D* eulerAngles) {
	return new QQuaternion(QQuaternion::fromEulerAngles(*eulerAngles));
}

void QQuaternion_GetEulerAngles(const QQuaternion* self, float* pitch, float* yaw, float* roll) {
	self->getEulerAngles(pitch, yaw, roll);
}

QQuaternion* QQuaternion_FromEulerAngles2(float pitch, float yaw, float roll) {
	return new QQuaternion(QQuaternion::fromEulerAngles(pitch, yaw, roll));
}

QGenericMatrix<3, 3, float> QQuaternion_ToRotationMatrix(const QQuaternion* self) {
	return self->toRotationMatrix();
}

QQuaternion* QQuaternion_FromRotationMatrix(const QGenericMatrix<3, 3, float>* rot3x3) {
	return new QQuaternion(QQuaternion::fromRotationMatrix(*rot3x3));
}

void QQuaternion_GetAxes(const QQuaternion* self, QVector3D* xAxis, QVector3D* yAxis, QVector3D* zAxis) {
	self->getAxes(xAxis, yAxis, zAxis);
}

QQuaternion* QQuaternion_FromAxes(const QVector3D* xAxis, const QVector3D* yAxis, const QVector3D* zAxis) {
	return new QQuaternion(QQuaternion::fromAxes(*xAxis, *yAxis, *zAxis));
}

QQuaternion* QQuaternion_FromDirection(const QVector3D* direction, const QVector3D* up) {
	return new QQuaternion(QQuaternion::fromDirection(*direction, *up));
}

QQuaternion* QQuaternion_RotationTo(const QVector3D* from, const QVector3D* to) {
	return new QQuaternion(QQuaternion::rotationTo(*from, *to));
}

QQuaternion* QQuaternion_Slerp(const QQuaternion* q1, const QQuaternion* q2, float t) {
	return new QQuaternion(QQuaternion::slerp(*q1, *q2, t));
}

QQuaternion* QQuaternion_Nlerp(const QQuaternion* q1, const QQuaternion* q2, float t) {
	return new QQuaternion(QQuaternion::nlerp(*q1, *q2, t));
}

void QQuaternion_Delete(QQuaternion* self) {
    delete self;
}

